include/mata/utils/ord-vector.hh

```diff
@@ -155,52 +162,63 @@ public:
 
         reserve_on_insert(vec_);
 
-        // perform binary search (cannot use std::binary_search because it is
+        // A conversation between someone and someone else:
+        // Someone: perform binary search (cannot use std::binary_search because it is
         // ineffective due to not returning the iterator to the position of the
         // desirable insertion in case the searched element is not present in the
         // range)
-        size_t first = 0;
-        size_t last = vec_.size();
-
-        if ((last != 0) && (vec_.back() < x))
-        {	// for the case which would be prevalent
-            // that is, the added thing can is larger than the largest thing and can be just bushed back
-            vec_.push_back(x);
-            return;
-        }
-
-        while (first < last)
-        {	// while the pointers do not overlap
-            size_t middle = first + (last - first) / 2;
-            if (vec_[middle] == x)
-            {	// in case we found x
-                return;
-            }
-            else if (vec_[middle] < x)
-            {	// in case middle is less than x
-                first = middle + 1;
-            }
-            else
-            {	// in case middle is greater than x
-                last = middle;
-            }
-        }
-
-        vec_.resize(vec_.size() + 1);
-        std::copy_backward(vec_.begin() + static_cast<long>(first), vec_.end() - 1, vec_.end());
-
-        // insert the new element
-        vec_[first] = x;
+        // Someone else: Or we can use lower_bound like this, and remove all these comments and code? Is there a problem (except the one additional comparison)?
+        auto pos  = std::lower_bound(vec_.begin(), vec_.end(), x);
```

Comment:
On side note, quite recently, I have read interesting article about very efficient binary search: https://mhdm.dev/posts/sb_lower_bound/. Not sure if it is applicable or it would even help things out, but if you want to squeeze some more juice, maybe we could check it out.

Reply:
It is an interesting article. Maybe we should put some effort into it. It is called so often that even a small improvement may have a great impact.

Reply:
I will put a note with the link to the source for now.

Reply:
We should create an issue for applying the information from the article. Otherwise, we will completely forget about this.

Reply:
done